### PR TITLE
Added Pagination Config Option Through Headers.

### PR DESCRIPTION
**Pull request by nearbycoder**
This will allow for pagination information to be past through the
headers on both the request and response. The request should have a
Range-Unit: items and a Content-Range: 0-4. The 0-4 is the amount of
items per page via the current page.

In addition the response will return the Content-Range: 0-4/10, the 0-4
is the range you requested and the 10 will be the total posts.

This change pairs well with the
angular-paginate-anything library found here:
https://github.com/begriffs/angular-paginate-anything

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -214,16 +215,17 @@ GEM
...
     slop (3.6.0)
-    sprockets (3.4.0)
+    sprockets (3.5.2)
+      concurrent-ruby (~> 1.0)
       rack (> 1, < 3)
-    sprockets-rails (2.3.3)
-      actionpack (>= 3.0)
-      activesupport (>= 3.0)
-      sprockets (>= 2.8, < 4.0)
+    sprockets-rails (3.0.4)
```
Comment (claycarpenter) [inline Gemfile.lock:221]: I have concerns about the downstream effects of these major version updates. Where the dependency updates necessary?

**Comment by claycarpenter** [comment]
This new configuration option probably should also be noted in an update to the project's README.

File: app/controllers/fuel_controller.rb [hunk 2]
```diff
@@ -14,4 +14,44 @@ def s3_bucket
     )
   end
 
+  def self.paginated_action(options={})
+    before_filter(options) do |controller|
```
Comment (claycarpenter) [inline app/controllers/fuel_controller.rb:18]: Suggest determining all pagination attributes here.

Look for the `config.header_pagination` flag, and if present then attempt to use Range. If not present then look for `params[:page]`. If all pagination configs are left undefined then set current page to `1` and per_page to `nil`; I believe that instructs Kaminari to return everything.

File: app/controllers/fuel_controller.rb [hunk 3]
```diff
@@ -14,4 +14,44 @@ def s3_bucket
...
+      # Set defaults
+      @pagination_current_page = 1
+      @pagination_per_page = nil
+
+      if request.headers['Range-Unit'] == 'items' && request.headers['Range'].present?
```
Comment (claycarpenter) [inline app/controllers/fuel_controller.rb]: How does these Range options interact with the existing `config.paginates_per` configuration?

File: app/controllers/fuel/posts_controller.rb [hunk 4]
```diff
@@ -3,13 +3,20 @@ class PostsController < FuelController
     include ActionView::Helpers::TextHelper
     layout Fuel.configuration.layout if Fuel.configuration.layout
     before_filter :define_title
+    if Fuel.configuration.header_pagination
+      paginated_action only: [:index]
```
Comment (claycarpenter) [inline app/controllers/fuel/posts_controller.rb]: Suggest treating all calls to PostsController#index as pagination-aware calls. See my comments on `fuel_controller.rb`.

File: app/controllers/fuel/posts_controller.rb [hunk 5]
```diff
@@ -3,13 +3,20 @@ class PostsController < FuelController
...
+    end
 
     def define_title
       @blog_title = Fuel.configuration.blog_title
     end
 
     def index
-      @posts = Fuel::Post.recent_published_posts.page(params[:page])
+      if Fuel.configuration.header_pagination
+        @posts = Fuel::Post.recent_published_posts.page(@pagination_current_page).per(@pagination_per_page)
```
Comment (claycarpenter) [inline app/controllers/fuel/posts_controller.rb]: We should be able to determine enough during the pre-action filter (see my comments on `fuel_controller.rb`) that we can get away with using only this single line to retrieve Posts for all invocations of the action.

File: app/controllers/fuel_controller.rb (hunk 2 above)
Comment (nearbycoder) [inline app/controllers/fuel_controller.rb:18]: Good thinking. Less checking within the posts controller.

File: app/controllers/fuel_controller.rb (hunk 3 above)
Comment (nearbycoder) [inline app/controllers/fuel_controller.rb]: It directly counteracts any changes to paginates_per, so it is either one or the other.

**Comment by nearbycoder** [comment]
Ready to be reviewed.

File: app/controllers/fuel_controller.rb [hunk 6]
```diff
@@ -14,4 +14,44 @@ def s3_bucket
     )
   end
 
+  def self.paginated_action(options={})
+    before_filter(options) do |controller|
+
+      # Set defaults
+      @pagination_current_page = params[:page] ? params[:page] : 1
+      @pagination_per_page = nil
```
Comment (claycarpenter) [inline app/controllers/fuel_controller.rb:22]: Perhaps I'm reading this wrong, but shouldn't this `pagination_per_page` var be set to `config.paginates_per` when not using the `header_pagination`-style pagination?

**Comment by claycarpenter** [comment]
Looks good :+1:

**Pull request merged**